packages/maskbook/src/plugins/Voicechat/utils/voicechat.ts

```diff
@@ -0,0 +1,347 @@
+import io from 'socket.io-client'
+import { ICE_SERVERS, SIGNALING_SERVER } from '../constants'
+import type { PeerMediaElement } from '../types'
+import { VolumeMeter } from './volumeMeter'
+
+let signalingSocket: SocketIOClient.Socket | null = null
+let localMediaStream: MediaStream | null = null
+let peers: Record<string, RTCPeerConnection> = {}
+let peerMediaElements: Record<string, PeerMediaElement> = {}
+let localPeerMediaElement: PeerMediaElement | null = null
+let speakingUsers: string[] = []
+let activeMediaDevice: string = ''
+
+let audioContext: AudioContext | null = null
+
+type StateCallback = (a: Record<string, PeerMediaElement>, b: PeerMediaElement | null) => void
+let stateCallback: StateCallback = (a, b) => {}
+let speakingCallback = (a: string[]) => {}
+let activeMediaDeviceCallback = (a: string) => {}
+
+export function init(
+    customServer: string,
+    id: string,
+    username: string,
+    callback: StateCallback,
+    activeMediaCallback: (a: string) => void,
+) {
+    // Do a disconnect
+    disconnectVoice()
+
+    // Register callbacks
+    stateCallback = callback
+    activeMediaDeviceCallback = activeMediaCallback
+
+    if (!audioContext) {
+        audioContext = new AudioContext()
+    }
+
+    console.log('connecting to signaling server')
+    signalingSocket = io(customServer ? customServer : SIGNALING_SERVER, { transports: ['websocket'] })
+
+    signalingSocket.on('connect', function () {
+        console.log('connected to signaling server')
+
+        setup_local_media(
+            () => {
+                join_chat_channel(id, { username })
+            },
+            () => {},
+            username,
+        )
+    })
+
+    setTimeout(() => {
+        if (!signalingSocket?.connected) {
+            alert('Could not connect to voice server')
```

Comment:
Do we really need to popup an alert here?

Reply:
i don't known